test/issue_746_helper.c

```diff
@@ -0,0 +1,383 @@
+		nextoff = telldir(d->dp);
+#ifdef __FreeBSD__		
+		/* Under FreeBSD, telldir() may return 0 the first time
+		   it is called. But for libfuse, an offset of zero
+		   means that offsets are not supported, so we shift
+		   everything by one. */
+		nextoff++;
+#endif
+		if (filler(buf, d->entry->d_name, &st, nextoff, fill_flags))
+			break;
+
+		d->entry = NULL;
+		d->offset = nextoff;
+	}
+
+	return 0;
+}
+
+static int xmp_releasedir(const char *path, struct fuse_file_info *fi)
+{
+	struct xmp_dirp *d = get_dirp(fi);
+	(void) path;
+	closedir(d->dp);
+	free(d);
+	return 0;
+}
+
+static int xmp_mkdir(const char *path, mode_t mode)
+{
+	int res;
+
+	res = mkdir(path, mode);
+	if (res == -1)
+		return -errno;
+
+	return 0;
+}
+
+static int xmp_unlink(const char *path)
+{
+	int res;
+
+	res = unlink(path);
+	if (res == -1)
+		return -errno;
+
+	return 0;
+}
+
+static int xmp_rmdir(const char *path)
+{
+	int res;
+
+	res = rmdir(path);
+	if (res == -1)
+		return -errno;
+
+	return 0;
+}
+
+static int xmp_rename(const char *from, const char *to, unsigned int flags)
+{
+	int res;
+
+	/* When we have renameat2() in libc, then we can implement flags */
+	if (flags)
+		return -EINVAL;
+
+        if(!getenv("TEST746_DELAY_DISABLE")) usleep(100000);
+
+	res = rename(from, to);
+	if (res == -1)
+		return -errno;
+
+	return 0;
+}
+
+static int xmp_chmod(const char *path, mode_t mode,
+		     struct fuse_file_info *fi)
+{
+	int res;
+
+	if(fi)
+		res = fchmod(fi->fh, mode);
+	else
+		res = chmod(path, mode);
+	if (res == -1)
+		return -errno;
+
+	return 0;
+}
+
+static int xmp_chown(const char *path, uid_t uid, gid_t gid,
+		     struct fuse_file_info *fi)
+{
+	int res;
+
+	if (fi)
+		res = fchown(fi->fh, uid, gid);
+	else
+		res = lchown(path, uid, gid);
+	if (res == -1)
+		return -errno;
+
+	return 0;
+}
+
+static int xmp_truncate(const char *path, off_t size,
+			 struct fuse_file_info *fi)
+{
+	int res;
+
+	if(fi)
+		res = ftruncate(fi->fh, size);
+	else
+		res = truncate(path, size);
+
+	if (res == -1)
+		return -errno;
+
+	return 0;
+}
+
+static int xmp_create(const char *path, mode_t mode, struct fuse_file_info *fi)
+{
+	int fd;
+
+	fd = open(path, fi->flags, mode);
+	if (fd == -1)
+		return -errno;
+
+	fi->fh = fd;
+	return 0;
+}
+
+static int xmp_open(const char *path, struct fuse_file_info *fi)
+{
+	int fd;
+
+	fd = open(path, fi->flags);
+	if (fd == -1)
+		return -errno;
+
+	fi->fh = fd;
+	return 0;
+}
+
+static int xmp_read(const char *path, char *buf, size_t size, off_t offset,
+		    struct fuse_file_info *fi)
+{
+	int res;
+
+	(void) path;
+	res = pread(fi->fh, buf, size, offset);
+	if (res == -1)
+		res = -errno;
+
+	return res;
+}
+
+static int xmp_write(const char *path, const char *buf, size_t size,
+		     off_t offset, struct fuse_file_info *fi)
+{
+	int res;
+
+	(void) path;
+	res = pwrite(fi->fh, buf, size, offset);
+	if (res == -1)
+		res = -errno;
+
+	return res;
+}
+
+static int xmp_statfs(const char *path, struct statvfs *stbuf)
```

Comment:
Do we need to implement all this functionality to make a minimal reproduction for a test here?

Reply:
If we keep it, it could be used for other tests?

Reply:
I don't know if it is needed for the pytest mount/unmount/cleanup logic in the CI etc.

Reply:
If we want to use it for other tests, I suggest folding it back into the general passthrough example, perhaps?  Or otherwise giving it a more descriptive name?

Reply:
Idea was to keep it in tests, to avoid polluting example/ with test specific things.

Reply:
Yes there is an earlier lengthy discussion about this. Due to the nature of the bug either you have to keep the existing example and add a PRELOAD library, or taint the example with inline test code, or create and maintain a new dedicated FUSE binary.

Reply:
Oh, I see.  Yes keeping the examples simple is indeed a good idea.

Reply:
Due to the nature of the bug and the type of test, you need at least these:

```
        .init           = xmp_init,
        .getattr        = xmp_getattr,
        .access         = xmp_access,
        .opendir        = xmp_opendir,
        .readdir        = xmp_readdir,
        .releasedir     = xmp_releasedir,
        .mkdir          = xmp_mkdir,
        .unlink         = xmp_unlink,
        .rmdir          = xmp_rmdir,
        .rename         = xmp_rename,
        .create         = xmp_create,
        .release        = xmp_release,
```

Could remove `rmdir` if you want to leave the temporary directory behind (not sure if `TemporaryDirectory` likes that).